kernel/dev/serial/rs232.cpp

```diff
@@ -48,11 +49,18 @@ void px_rs232_print(char* str) {
 }
 
 static void px_rs232_callback(registers_t *regs) {
-    char str[2] = { px_rs232_read_char(), '\0' };
-    if (str[0] == '\r') {
-        px_rs232_print("\n");
+    // Grab the input character
+    char in = px_rs232_read_char();
+    // Change carriage returns to newlines
+    if (in == '\r') {
+        in = '\n';
     }
+    // Create a string and print it so that the
+    // user can see what they're typing.
+    char str[2] = {in, '\0'};
     px_rs232_print(str);
+    // Add the character to the circular buffer
+    px_ring_buffer_enqueue(read_buffer, (uint8_t)in);
```

Comment:
What happens if someone sends serial data before the buffer is initialized? Do we just wait to enable interrupts until after both have been initialized?

Reply:
I'm fairly certain the input is just ignored.

Reply:
So no null dereference? That's what it looks like would happen.

Reply:
This code is pretty similar to what it was before, and we didn't get null dereferences then, but I can double check that later.

Reply:
I guess the ring buffer code should be checking that. It just doesn't feel right to me.